### PR TITLE
Step 10: Address, Delivery, Payment

### PR DESCRIPTION
**Pull request by vtodorchuk**

### CODE REVIEW

File: app/jobs/cart_cleaner_job.rb [hunk 1]
```diff
@@ -1,6 +1,7 @@
-class CartCleaner
+class CartCleanerJob
```
Comment (CaptainJNS) [inline app/jobs/cart_cleaner_job.rb]: it's better to name `CartCleanerWorker` and folder name also is `workers`

File: app/decorators/delivery_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class DeliveryDecorator < Draper::Decorator
+  delegate_all
+
+  def price
+    Money.new(price_cents)
+  end
```
Comment (CaptainJNS) [inline app/decorators/delivery_decorator.rb:6]: You don't need it if you use gem `money`. It already creates a `price` helper method for field `price_cents`

File: app/services/checkout_service.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def self.count_total_price(order)
+    order_book = OrderBook.where(order_id: order.id)
+
+    total_price = order_book.map do |product|
+      product.book.price_cents * product.quantity
+    end.sum
+
+    total_price *= order.coupon.discount unless order.coupon.nil?
```
Comment (CaptainJNS) [inline app/services/checkout_service.rb]:
```suggestion
    total_price *= order.coupon.discount if order.coupon.present?
```

File: app/services/checkout_service.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
+class CheckoutService
+  def self.to_errors(message)
+    message.map do |k, v|
```
Comment (CaptainJNS) [inline app/services/checkout_service.rb]:
```suggestion
    message.map do |key, value|
```

**Review by CaptainJNS — APPROVED** [review]

File: .circleci/config.yml [hunk 5]
```diff
@@ -43,7 +43,7 @@ commands:
             - vendor/bundle
       - run:
           name: Install dependencies
-          command: bundle check --path=vendor/bundle || bundle install --path=vendor/bundle --jobs 4 --retry 3
+          command: bundle check --path=vendor/bundle || bundle install --path=vendor/bundle --workers 4 --retry 3
```
Comment (kolyanjy) [inline .circleci/config.yml:46]: retry is bad practice

File: app/controllers/checkouts_controller.rb [hunk 6]
```diff
@@ -1,16 +1,123 @@
...
+
+  def allow_step
+    order = Order.find(session[:order_id])
+
+    if order.books.empty?
+      redirect_to books_path
+    elsif %w[confirm complete].include?(permitted_params[:step]) && (order.address.nil? || order.delivery.nil?)
+      redirect_to checkout_path(step: :address)
+    end
+  end
```
Comment (kolyanjy) [inline app/controllers/checkouts_controller.rb:112]: should be in separate service objects

File: app/controllers/orders_controller.rb [hunk 7]
```diff
@@ -1,14 +1,63 @@
...
+    end
+  end
+
+  def update_user_address(permitted_params)
+    if permitted_params[:shipping].nil?
+      current_user.billing.update(permitted_params[:billing])
     else
-      redirect_to new_checkout_path
+      current_user.shipping.update(permitted_params[:shipping])
     end
```
Comment (kolyanjy) [inline app/controllers/orders_controller.rb:59]: the same as above

File: app/services/checkout_service.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def self.check_fields(permitted_params)
+    case permitted_params[:step]
+    when 'shipping'
+      check(permitted_params[:shipping]) unless permitted_params[:shipping].nil?
+    when 'billing'
+      check(permitted_params[:billing]) unless permitted_params[:billing].nil?
+    when 'payment'
+      check(permitted_params[:credit_card]) unless permitted_params[:credit_card].nil?
```
Comment (kolyanjy) [inline app/services/checkout_service.rb:35]: better to rewrite this service with instance methods

**Pull request merged**